machineconfiguration/v1/types_machineosbuild.go

```diff
@@ -0,0 +1,208 @@
+// +kubebuilder:printcolumn:name="Failed",type="string",JSONPath=.status.conditions[?(@.type=="Failed")].status
+
+// MachineOSBuild describes a build process managed and deployed by the MCO
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSBuild struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machine os build
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="machineOSBuildSpec is immutable once set"
+	// +required
+	Spec MachineOSBuildSpec `json:"spec"`
+
+	// status describes the last observed state of this machine os build
+	// +optional
+	Status *MachineOSBuildStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineOSBuildList describes all of the Builds on the system
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	// items contains a collection of MachineOSBuild resources.
+	// +optional
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes information about a build process primarily populated from a MachineOSConfig object.
+type MachineOSBuildSpec struct {
+	// machineConfig points to the rendered MachineConfig resource to be included in this image build.
+	// +required
+	MachineConfig MachineConfigReference `json:"machineConfig"`
+	// machineOSConfig references the MachineOSConfig resource that this image build extends.
+	// +required
+	MachineOSConfig MachineOSConfigReference `json:"machineOSConfig"`
+	// renderedImagePushSpec is set by the Machine Config Operator from the MachineOSConfig object this build is attached to.
+	// This field describes the location of the final image, which will be pushed by the build once complete.
+	// The format of the image push spec is: host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>.
+	// The length of the push spec must be between 1 to 447 characters.
+	// +required
+	RenderedImagePushSpec ImageTagFormat `json:"renderedImagePushSpec"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+// +kubebuilder:validation:XValidation:rule="has(self.buildEnd) ? has(self.buildStart) && timestamp(self.buildStart) < timestamp(self.buildEnd) : true",message="buildEnd must be after buildStart"
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, and Succeeded.
+	// Once a Build is marked as Failed or Interrupted, no future conditions can be set.
```

Comment:
So, question on the mechanics of this, does the controller just see these states and bail out early? There's no point where it's trying to update the Failed or Interrupted conditions even after they've been added?

Reply:
The controller is the only updator, and prevents itself from resetting the status once in one of the terminal states:

https://github.com/openshift/machine-config-operator/blob/67789532e4569ece95eb98fed1dde03409f065e9/pkg/controller/build/helpers.go#L162

I'll also add "success" to the validations here since you shouldn't be able to transition out of it.

cc @cheesesashimi in case I'm missing something

Reply:
That's correct. Once the MachineOSBuild reaches a terminal state (e.g., success, failed, interrupted), it won't try to do any other updates on it.